File: src/ios/NativeAudioAsset.m
//
//
//  NativeAudioAsset.m
//  NativeAudioAsset
//
//  Created by Sidney Bofah on 2014-06-26.
//

#import "NativeAudioAsset.h"

@interface NativeAudioAsset () <AVAssetResourceLoaderDelegate>

@property (nonatomic, strong) AVPlayer *player;

@end

@implementation NativeAudioAsset


static const CGFloat FADE_STEP = 0.05;
static const CGFloat FADE_DELAY = 0.08;

-(id) initWithPath:(NSString*) path withVolume:(NSNumber*) volume withFadeDelay:(NSNumber *)delay withControlsInfo:(NSDictionary *) controlsInfo
{
    setenv("CFNETWORK_DIAGNOSTICS","3",1);
    self = [super init];
    if(self) {
        
        NSURL *pathURL;
        
        if([path hasPrefix:@"http"]){
            pathURL = [NSURL URLWithString: path];
        }else{
            pathURL = [NSURL fileURLWithPath: path];
        }
        
        if(controlsInfo != nil){
            self->controlsInfo = [[MusicControlsInfo alloc] initWithDictionary:controlsInfo];
        } else {
            self->controlsInfo = nil;
        }
        
        AVURLAsset *asset = [AVURLAsset URLAssetWithURL:pathURL options:nil];
        [asset.resourceLoader setDelegate:self queue:dispatch_get_main_queue()];
        AVPlayerItem * playerItem = [AVPlayerItem playerItemWithAsset:asset automaticallyLoadedAssetKeys:@[@"playable",@"duration"]];
        self.player = [[AVPlayer alloc] initWithPlayerItem:playerItem];
        
        [self.player addObserver:self forKeyPath:@"status" options:NSKeyValueObservingOptionNew context:NULL];
        [self.player addObserver:self forKeyPath:@"timeControlStatus" options:NSKeyValueObservingOptionNew context:NULL];
        self.player.volume = volume.floatValue;
        
        if(delay)
        {
            fadeDelay = delay;
        }
        else {
            fadeDelay = [NSNumber numberWithFloat:FADE_DELAY];
        }
        
        initialVolume = volume;
       
        
    }
    return(self);
}

- (void)addObservers
{
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(playerItemDidReachEnd:)
                                                 name:AVPlayerItemDidPlayToEndTimeNotification
                                               object:[self.player currentItem]];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(playerItemDidStall:)
                                                 name:AVPlayerItemPlaybackStalledNotification
                                               object:[self.player currentItem]];
}


- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    
    if (object == self.player && [keyPath isEqualToString:@"status"]) {
        if (self.player.status == AVPlayerStatusFailed) {
            NSLog(@"AVPlayer Failed");
            
        } else if (self.player.status == AVPlayerStatusReadyToPlay) {
            NSLog(@"AVPlayerStatusReadyToPlay");
            //[player play];
            
        } else if (self.player.status == AVPlayerItemStatusUnknown) {
            NSLog(@"AVPlayer Unknown");
            
        }
    }
    
    if(object == self.player && [keyPath isEqualToString:@"timeControlStatus"]){
        if (self.player.timeControlStatus == AVPlayerTimeControlStatusPlaying) {
            NSLog(@"AVPlayerTimeControlStatusPlaying");
            if(self->playPauseCallback != nil) {
                self->playPauseCallback(self->audioId, true);
            }
            
        } else if (self.player.timeControlStatus == AVPlayerTimeControlStatusWaitingToPlayAtSpecifiedRate) {
            NSLog(@"AVPlayerTimeControlStatusWaitingToPlayAtSpecifiedRate");
            if(self->playPauseCallback != nil) {
                self->playPauseCallback(self->audioId, true);
            }
            
        } else if (self.player.timeControlStatus == AVPlayerTimeControlStatusPaused) {
            NSLog(@"AVPlayerTimeControlStatusPaused");
            if(self->playPauseCallback != nil) {
                self->playPauseCallback(self->audioId, false);
            }
            
        }
    }
}


- (void) play
{
    [self.player play];
}


// The volume is increased repeatedly by the fade step amount until the last step where the audio is stopped.
// The delay determines how fast the decrease happens
- (void)playWithFade
{
    //rate=0 -> not playing
    if (self.player.rate == 0)
    {
        self.player.volume = 0;
        [self.player play];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self performSelector:@selector(playWithFade) withObject:nil afterDelay:self->fadeDelay.floatValue];
        });
    }
    else
    {
        if(self.player.volume < initialVolume.floatValue)
        {
            self.player.volume += FADE_STEP;
            dispatch_async(dispatch_get_main_queue(), ^{
                [self performSelector:@selector(playWithFade) withObject:nil afterDelay:self->fadeDelay.floatValue];
            });
        }
    }
}

- (void) pause
{
    [self.player pause];
}

// The volume is decreased repeatedly by the fade step amount until the volume reaches the configured level.
// The delay determines how fast the increase happens
- (void)pauseWithFade
{
    BOOL shouldContinue = NO;
    
    //rate=0 -> not playing
    if (self.player.rate != 0 && self.player.volume > FADE_STEP) {
        self.player.volume -= FADE_STEP;
        shouldContinue = YES;
    } else {
        // Pause and get the sound ready for playing again
        [self.player pause];
        self.player.volume = initialVolume.floatValue;
    }
    
    if(shouldContinue) {
        [self performSelector:@selector(pauseWithFade) withObject:nil afterDelay:fadeDelay.floatValue];
    }
}

- (void) stop
{
    //There is no stop method in AVPlayer, so we pause and go to position 0
    
    [self.player pause];
    [self.player seekToTime:CMTimeMake(0, 1)];
    
}

// The volume is decreased repeatedly by the fade step amount until the volume reaches the configured level.
// The delay determines how fast the increase happens
- (void)stopWithFade
{
    BOOL shouldContinue = NO;
    
    //rate=0 -> not playing
    if (self.player.rate != 0 && self.player.volume > FADE_STEP) {
        self.player.volume -= FADE_STEP;
        shouldContinue = YES;
    } else {
        // Stop and get the sound ready for playing again
        //There is no stop method in AVPlayer, so we pause and go to position 0
        [self.player pause];
        self.player.volume = initialVolume.floatValue;
        [self.player seekToTime:CMTimeMake(0, 1)];
    }
    
    
    if(shouldContinue) {
        [self performSelector:@selector(stopWithFade) withObject:nil afterDelay:fadeDelay.floatValue];
    }
}

- (void) loop
{
    [self stop];
    
    [self.player seekToTime:CMTimeMake(0, 1)];
    
    self.player.actionAtItemEnd = AVPlayerActionAtItemEndNone;
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(playerItemDidReachEnd:)
                                                 name:AVPlayerItemDidPlayToEndTimeNotification
                                               object:[self.player currentItem]];
    
    [self.player play];
}

- (void)playerItemDidReachEnd:(NSNotification *)notification {
    [self.player seekToTime:CMTimeMake(0, 1)];
}

- (void)playerItemDidStall:(NSNotification *)notification
{
    //Do nothing for now
}

- (void) unload
{
    [self stop];
    
    self.player = nil;
}

- (void) setVolume:(NSNumber*) volume;
{
    
    [self.player setVolume:volume.floatValue];
    
}

- (void) setCallbackAndId:(CompleteCallback)cb audioId:(NSString*)aID
{
    self->audioId = aID;
    self->finished = cb;
}

- (void) setPlayPauseCallbackAndId:(PlayPauseCallback)cb audioId:(NSString*)aID
{
    self->audioId = aID;
    self->playPauseCallback = cb;
    
}

- (void) audioPlayerDidFinishPlaying:(AVPlayer *)player successfully:(BOOL)flag
{
    if (self->finished) {
        self->finished(self->audioId);
    }
}

- (void) audioPlayerDecodeErrorDidOccur:(AVPlayer *)player error:(NSError *)error
{
    if (self->finished) {
        self->finished(self->audioId);
    }
}

- (NSTimeInterval) getDuration;
{
    
    //AVPlayerItem is not ready to give us the value, so we return 0
    if(self.player.status != AVPlayerItemStatusReadyToPlay){
        NSLog(@"NOT READY TO PLAY");
        return 0;
    }if(CMTIME_IS_INDEFINITE(self.player.currentItem.duration)){
        return 0;
    }else{
        return CMTimeGetSeconds(self.player.currentItem.duration)*1000;
    }
}

- (NSTimeInterval) getCurrentPosition;
{
    return CMTimeGetSeconds(self.player.currentTime)*1000;
}

- (void) seekTo:(NSNumber*)position;
{
    [self.player seekToTime:CMTimeMake(position.intValue, 1000)];
}

- (NSString*) getTrackName;
{
    if(self->controlsInfo){
        return [self->controlsInfo track];
    } else {
        return nil;
    }
}

- (void) skipForward;
{
    [self.player seekToTime: CMTimeAdd(self.player.currentTime,CMTimeMakeWithSeconds(10, 1))];
}


- (void) skipBackward;
{
    [self.player seekToTime: CMTimeSubtract(self.player.currentTime,CMTimeMakeWithSeconds(10, 1))];
}

- (MusicControlsInfo *) getControlsInfo;
{
    return self->controlsInfo;
}

@end

File: src/ios/NativeAudioAsset.swift
//  Converted to Swift 5.3 by Swiftify v5.3.19197 - https://swiftify.com/
//
//
//  NativeAudioAsset.swift
//  NativeAudioAsset
//
//  Created by Sidney Bofah on 2014-06-26.
//

let FADE_STEP: CGFloat = 0.05
let FADE_DELAY: CGFloat = 0.08

class NativeAudioAsset: AVAssetResourceLoaderDelegate {
    private var player: AVPlayer?

    init(path: String?, withVolume volume: NSNumber?, withFadeDelay delay: NSNumber?, withTrackName name: String?) {
        setenv("CFNETWORK_DIAGNOSTICS", "3", 1)
        super.init()
        var pathURL: URL?

        if path?.hasPrefix("http") ?? false {
            pathURL = URL(string: path ?? "")
        } else {
            pathURL = URL(fileURLWithPath: path ?? "")
        }

        var asset: AVURLAsset? = nil
        if let pathURL = pathURL {
            asset = AVURLAsset(url: pathURL, options: nil)
        }
        asset?.resourceLoader.setDelegate(self, queue: DispatchQueue.main)
        var playerItem: AVPlayerItem? = nil
        if let asset = asset {
            playerItem = AVPlayerItem(asset: asset, automaticallyLoadedAssetKeys: ["playable", "duration"])
        }
        player = AVPlayer(playerItem: playerItem)

        player?.addObserver(self, forKeyPath: "status", options: .new, context: nil)
        player?.addObserver(self, forKeyPath: "timeControlStatus", options: .new, context: nil)
        player?.volume = volume?.floatValue ?? 0.0

        if let delay = delay {
            fadeDelay = delay
        } else {
            fadeDelay = NSNumber(value: Float(FADE_DELAY))
        }

        initialVolume = volume

        if let name = name {
            trackName = name
        } else {
            trackName = nil
        }
        self
    }

    func addObservers() {

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerItemDidReachEnd(_:)),
            name: .AVPlayerItemDidPlayToEndTime,
            object: player?.currentItem)
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerItemDidStall(_:)),
            name: .AVPlayerItemPlaybackStalled,
            object: player?.currentItem)
    }

    func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [String : Any]?, context: UnsafeMutableRawPointer?) {

        if (object as? AVPlayer) == player && (keyPath == "status") {
            if player?.status == .failed {
                print("AVPlayer Failed")
            } else if player?.status == .readyToPlay {
                print("AVPlayerStatusReadyToPlay")
                //[player play];
            } else if player?.status == AVPlayerItem.Status.unknown {
                print("AVPlayer Unknown")
            }
        }

        if (object as? AVPlayer) == player && (keyPath == "timeControlStatus") {
            if player?.timeControlStatus == .playing {
                print("AVPlayerTimeControlStatusPlaying")
                if let playPauseCallback = playPauseCallback {
                    playPauseCallback(audioId, true)
                }
            } else if player?.timeControlStatus == .waitingToPlayAtSpecifiedRate {
                print("AVPlayerTimeControlStatusWaitingToPlayAtSpecifiedRate")
                if let playPauseCallback = playPauseCallback {
                    playPauseCallback(audioId, true)
                }
            } else if player?.timeControlStatus == .paused {
                print("AVPlayerTimeControlStatusPaused")
                if let playPauseCallback = playPauseCallback {
                    playPauseCallback(audioId, false)
                }
            }
        }
    }

    func play() {
        player?.play()
    }

    // The volume is increased repeatedly by the fade step amount until the last step where the audio is stopped.
    // The delay determines how fast the decrease happens
    func playWithFade() {
        //rate=0 -> not playing
        if player?.rate == 0 {
            player?.volume = 0
            player?.play()
            DispatchQueue.main.async(execute: { [self] in
                perform(#selector(playWithFade), with: nil, afterDelay: TimeInterval(fadeDelay.floatValue))
            })
        } else {
            if (player?.volume ?? 0.0) < initialVolume.floatValue {
                player?.volume += Float(FADE_STEP)
                DispatchQueue.main.async(execute: { [self] in
                    perform(#selector(playWithFade), with: nil, afterDelay: TimeInterval(fadeDelay.floatValue))
                })
            }
        }
    }

    func pause() {
        player?.pause()
    }

    // The volume is decreased repeatedly by the fade step amount until the volume reaches the configured level.
    // The delay determines how fast the increase happens
    func pauseWithFade() {
        var shouldContinue = false

        //rate=0 -> not playing
        if player?.rate != 0 && CGFloat(player?.volume ?? 0.0) > FADE_STEP {
            player?.volume -= Float(FADE_STEP)
            shouldContinue = true
        } else {
            // Pause and get the sound ready for playing again
            player?.pause()
            player?.volume = initialVolume.floatValue
        }

        if shouldContinue {
            perform(#selector(pauseWithFade), with: nil, afterDelay: TimeInterval(fadeDelay.floatValue))
        }
    }

    func stop() {
        //There is no stop method in AVPlayer, so we pause and go to position 0

        player?.pause()
        player?.seek(to: CMTimeMake(value: 0, timescale: 1))

    }

    // The volume is decreased repeatedly by the fade step amount until the volume reaches the configured level.
    // The delay determines how fast the increase happens
    func stopWithFade() {
        var shouldContinue = false

        //rate=0 -> not playing
        if player?.rate != 0 && CGFloat(player?.volume ?? 0.0) > FADE_STEP {
            player?.volume -= Float(FADE_STEP)
            shouldContinue = true
        } else {
            // Stop and get the sound ready for playing again
            //There is no stop method in AVPlayer, so we pause and go to position 0
            player?.pause()
            player?.volume = initialVolume.floatValue
            player?.seek(to: CMTimeMake(value: 0, timescale: 1))
        }


        if shouldContinue {
            perform(#selector(stopWithFade), with: nil, afterDelay: TimeInterval(fadeDelay.floatValue))
        }
    }

    func loop() {
        stop()

        player?.seek(to: CMTimeMake(value: 0, timescale: 1))

        player?.actionAtItemEnd = .none

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerItemDidReachEnd(_:)),
            name: .AVPlayerItemDidPlayToEndTime,
            object: player?.currentItem)

        player?.play()
    }

    @objc func playerItemDidReachEnd(_ notification: Notification?) {
        player?.seek(to: CMTimeMake(value: 0, timescale: 1))
    }

    @objc func playerItemDidStall(_ notification: Notification?) {
        //Do nothing for now
    }

    func unload() {
        stop()

        player = nil
    }

    func setVolume(_ volume: NSNumber?) {

        player?.volume = volume?.floatValue ?? 0.0

    }

    func setCallbackAndId(_ cb: CompleteCallback, audioId aID: String?) {
        audioId = aID
        finished = cb
    }

    func setPlayPauseCallbackAndId(_ cb: PlayPauseCallback, audioId aID: String?) {
        audioId = aID
        playPauseCallback = cb

    }

    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        if finished {
            finished(audioId)
        }
    }

    func audioPlayerDecodeErrorDidOccur(_ player: AVAudioPlayer, error: Error?) {
        if finished {
            finished(audioId)
        }
    }

    func getDuration() -> TimeInterval {

        //AVPlayerItem is not ready to give us the value, so we return 0
        if player?.status != AVPlayerItem.Status.readyToPlay {
            print("NOT READY TO PLAY")
            return 0
        }
        if CMTIME_IS_INDEFINITE(player?.currentItem?.duration) {
            return 0
        } else {
            if let duration = player?.currentItem?.duration {
                return TimeInterval(CMTimeGetSeconds(duration) * 1000)
            }
            return 0.0
        }
    }

    func getCurrentPosition() -> TimeInterval {
        if let currentTime = player?.currentTime() {
            return TimeInterval(CMTimeGetSeconds(currentTime) * 1000)
        }
        return 0.0
    }

    func seek(to position: NSNumber?) {
        player?.seek(to: CMTimeMake(value: Int64(position?.intValue ?? 0), timescale: 1000))
    }

    func getTrackName() -> String? {
        return trackName
    }

    func skipForward() {
        if let currentTime = player?.currentTime() {
            player?.seek(to: CMTimeAdd(currentTime, CMTimeMakeWithSeconds(Float64(10), preferredTimescale: 1)))
        }
    }

    func skipBackward() {
        if let currentTime = player?.currentTime() {
            player?.seek(to: CMTimeSubtract(currentTime, CMTimeMakeWithSeconds(Float64(10), preferredTimescale: 1)))
        }
    }
}